base/compiler/inferencestate.jl

```diff
@@ -1195,7 +1206,10 @@ function doworkloop(interp::AbstractInterpreter, sv::AbsIntState)
     completed = task(interp, sv)
     tasks = sv.tasks # allow dropping gc root over the previous call
     completed isa Bool || throw(TypeError(:return, "", Bool, task)) # print the task on failure as part of the error message, instead of just "@ workloop:line"
-    completed || push!(tasks, task)
+    if !completed
+        @assert length(tasks) >= prev "Task did not complete, but also did not create any child tasks"
+        push!(tasks, task)
+    end
```

Comment:
This introduces an artificial and unnecessary constraint, which I don't think is necessarily expected to be respected. In particular, occasionally a task might schedule a new frame, but not care about the result or might be implemented as handling the result itself.

Reply:
I don't think the constraint is arbitrary. If the task doesn't care about the result, it should mark itself as completed. The constraint is that if you did not complete, you better have scheduled some child tasks, because otherwise you're just getting rescheduled immediately, so at the very least it's an inefficiency to go back through the scheduler, but more likely you're about to infinite loop.

Reply:
That's only true of the inner task loop, but there also the cycle loop around this that handles entire frames. If the task knows or notices there is work to do on the cycle of frames, it could expect to wait for that work while not adding anything new itself. For example, infercalls and handle1 are currently split up in abstract_call_gf_by_type since it was easier to write that way, but they could have been merged into the same function using the state machine pattern.

Reply:
So amend the condition by also allowing something to be pushed onto `sv.callstack`. My problem is that it's way too easy to create an infinite loop here.

Reply:
Do you want to just update the assert to check if the callstack changed length here?